docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_webhook.go

```diff
@@ -0,0 +1,182 @@
+// +kubebuilder:webhook:path=/validate-batch-tutorial-kubebuilder-io-v1-cronjob,mutating=false,failurePolicy=fail,groups=batch.tutorial.kubebuilder.io,resources=cronjobs,verbs=create;update,versions=v1,name=vcronjob.kb.io
+
+/*
+We implement the `webhook.Defaulter` interface for our CronJob CRD.
+`webhook.Defaulter` only has one method: `Default()`.
+If `webhook.Defaulter` is implemented, a webhook server will be autowired to
+serve as the mutating webhook for CronJobs.
+*/
+
+var _ webhook.Defaulter = &CronJob{}
+
+// Default implements webhookutil.defaulter so a webhook will be registered for the type
+func (c *CronJob) Default() {
+	log.Info("defaulting cronjob", "namespace", c.Namespace, "name", c.Name)
+
+	if c.Spec.ConcurrencyPolicy == "" {
+		c.Spec.ConcurrencyPolicy = AllowConcurrent
+	}
+	if c.Spec.Suspend == nil {
+		c.Spec.Suspend = new(bool)
+	}
+	if c.Spec.SuccessfulJobsHistoryLimit == nil {
+		c.Spec.SuccessfulJobsHistoryLimit = new(int32)
+		*c.Spec.SuccessfulJobsHistoryLimit = 3
+	}
+	if c.Spec.FailedJobsHistoryLimit == nil {
+		c.Spec.FailedJobsHistoryLimit = new(int32)
+		*c.Spec.FailedJobsHistoryLimit = 1
+	}
+}
+
+/*
+We implement the `webhook.Validator` interface for our CronJob CRD.
+`webhook.Validator` interface has two methods: `ValidateCreate()` and `ValidateUpdate()`.
+If `webhook.Validator` is implemented, a webhook server will be autowired to
+serve as the validating webhook for CronJobs.
+*/
+
+var _ webhook.Validator = &CronJob{}
+
+// ValidateCreate implements webhookutil.validator so a webhook will be registered for the type
+func (c *CronJob) ValidateCreate() error {
+	log.Info("validate create", "namespace", c.Namespace, "name", c.Name)
+	return c.validateCronJob()
+}
+
+// ValidateUpdate implements webhookutil.validator so a webhook will be registered for the type
+func (c *CronJob) ValidateUpdate(old runtime.Object) error {
+	log.Info("validate update", "namespace", c.Namespace, "name", c.Name)
+	return c.validateCronJob()
+}
+
+/*
+We uses the same the validation logic for creation and update.
+*/
+
+func (c *CronJob) validateCronJob() error {
+	allErrs := c.validateCronJobSpec()
+	if len(c.ObjectMeta.Name) > apimachineryutilvalidation.DNS1035LabelMaxLength-11 {
+		// The cronjob controller appends a 11-character suffix to the cronjob (`-$TIMESTAMP`) when
+		// creating a job. The job name length limit is 63 characters.
+		// Therefore cronjob names must have length <= 63-11=52. If we don't validate this here,
+		// then job creation will fail later.
+		allErrs = append(allErrs, field.Invalid(field.NewPath("metadata").Child("name"), c.Name, "must be no more than 52 characters"))
+	}
+
+	return apierrors.NewInvalid(
+		schema.GroupKind{Group: "batch.tutorial.kubebuilder.io", Kind: "CronJob"},
+		c.Name, allErrs)
+}
+
+func (c *CronJob) validateCronJobSpec() field.ErrorList {
```

Comment:
add some comments in this.  Also, most of this isn't needed because we validate it declaratively.  We don't want to confuse users into thinking that both are needed.  We should make a point that you only really need this if you have complex logic that can't be validated declaratively (e.g. perhaps the format of the cron schedule?)